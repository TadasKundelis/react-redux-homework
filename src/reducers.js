import { combineReducers } from 'redux';

export const initialMoviesState = {
  list: [],
  hearted: [],
};

export const moviesReducer = (state = initialMoviesState, action) => {
  const { hearted } = state;
  switch (action.type) {
    case 'SET_MOVIE_LIST':
      // console.log(action.list);
      return { ...state, list: action.list };
    case 'ADD_TO_HEARTED':
      return { ...state, hearted: [...hearted, action.movie] };
    case 'REMOVE_FROM_HEARTED':
      return { ...state, hearted: hearted.filter(movie => movie !== action.movie) };
    default:
      return state;
  }
};


const initialGenresState = {
  list: [],
  selected: null,
};

export const genresReducer = (state = initialGenresState, action) => {
  switch (action.type) {
    case 'SET_GENRE_LIST':
      return { ...state, list: action.list };
    case 'SET_SELECTED_GENRE':
      return { ...state, selected: action.genre };
    default:
      return state;
  }
};

const initialLogState = [];

export const logsReducer = (state = initialLogState, action) => {
  switch (action.type) {
    case 'ADD_LOG':
      return [...state, action.log];
    default:
      return state;
  }
};

export default combineReducers({
  movies: moviesReducer,
  genres: genresReducer,
  logs: logsReducer,
});
