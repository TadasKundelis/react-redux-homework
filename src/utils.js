export const getFormatedDate = () => {
    const date = new Date();
    const [year, month, day, hour, minute, second] = [
        date.getFullYear(), 
        date.getMonth() + 1, 
        date.getDate(), 
        date.getHours(), 
        date.getMinutes(), 
        date.getSeconds()
    ].map(padDate);
    
    return `${year}-${month}-${day} ${hour}:${minute}:${second}: `;
}


const padDate = date => date < 10 ? `0${date}` : date;