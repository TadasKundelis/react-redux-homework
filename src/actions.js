import { getFormatedDate } from "./utils";
import { fetchGenresMovies } from "./thunks";

export const setMovieList = (list) => ({
  type: 'SET_MOVIE_LIST',
  list,
});


export const setGenreList = (list) => ({
  type: 'SET_GENRE_LIST',
  list
})

export const setSelectedGenre = (genre, genreName) => (dispatch, getState) => {
  if (getState().genres.selected === genre) return;
  const log = `Zanras pakeistas i ${genreName}`;
  dispatch(fetchGenresMovies(genre));
  dispatch(addLog(log));
  dispatch({
    type: 'SET_SELECTED_GENRE',
    genre
  })
}

export const addHeart = (movie, title) => dispatch => {
  const log = `${title} buvo uzsirdintas`;
  dispatch(addLog(log));
  dispatch ({
    type: 'ADD_TO_HEARTED',
    movie
  })
}

export const removeHeart = (movie, title) => dispatch => {
  const log = `${title} buvo nusirdintas`;
  dispatch(addLog(log));
  dispatch({
    type: 'REMOVE_FROM_HEARTED',
    movie
  })
}

export const appLoaded = (log) => dispatch => {
  dispatch(addLog(log))
}

export const addLog = (log) => ({
  type: "ADD_LOG",
  log: getFormatedDate() + log
})










