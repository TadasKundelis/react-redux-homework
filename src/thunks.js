import axios from 'axios';
import { endpoints } from '../config';
import { setMovieList, setGenreList, addLog } from './actions';

export const getPopularMovies = () => (dispatch) => {
  axios
    .get(endpoints.mostPopularMovies())
    .then((res) => dispatch(setMovieList(res.data.results)))
    .catch((error) => console.log(error));
};


export const fetchGenres = () => dispatch => {
  axios
    .get(endpoints.genres())
    .then(res => dispatch(setGenreList(res.data.genres)))
    .catch(error => console.log(error));
}


export const fetchGenresMovies = (genre) => (dispatch, getState) => {
  axios
    .get(endpoints.genreMovies(genre))
    .then((res) => dispatch(setMovieList(res.data.results)))
    .catch((error) => console.log(error))
}

