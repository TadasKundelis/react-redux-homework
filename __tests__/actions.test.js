import configureStore from 'redux-mock-store';
import * as actions from '../src/actions';
import { initialState } from '../src/reducers';

const mockStore = configureStore();
const store = mockStore(initialState);

describe('setMovieList action creator', () => {
  it('should create an action to set movie list', () => {
    store.dispatch(actions.setMovieList([]));
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('setGenreList action creator', () => {
  it('should create an action to set genre list', () => {
    store.dispatch(actions.setGenreList([]));
    expect(store.getActions()).toMatchSnapshot();
  });
});
