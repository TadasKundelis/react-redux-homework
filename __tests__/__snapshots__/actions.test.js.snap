// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addLog action creator should create an action to add a log 1`] = `
Array [
  Object {
    "list": Array [],
    "type": "SET_MOVIE_LIST",
  },
  Object {
    "list": Array [],
    "type": "SET_GENRE_LIST",
  },
  Object {
    "log": "2018-12-29 03:17:38: this is a new log",
    "type": "ADD_LOG",
  },
]
`;

exports[`setGenreList action creator should create an action to set genre list 1`] = `
Array [
  Object {
    "list": Array [],
    "type": "SET_MOVIE_LIST",
  },
  Object {
    "list": Array [],
    "type": "SET_GENRE_LIST",
  },
]
`;

exports[`setMovieList action creator should create an action to set movie list 1`] = `
Array [
  Object {
    "list": Array [],
    "type": "SET_MOVIE_LIST",
  },
]
`;
